subsys/modem/modem_ppp.c

```diff
@@ -0,0 +1,618 @@
+		modem_ppp_tx_net_pkt_buf_get(ppp, &ppp->tx_pkt);
+	}
+
+	if (ppp->tx_pkt != NULL) {
+		/* Initialize wrap */
+		if (ppp->transmit_state == MODEM_PPP_TRANSMIT_STATE_IDLE) {
+			ppp->transmit_state = MODEM_PPP_TRANSMIT_STATE_SOF;
+		}
+
+		/* Fill transmit ring buffer */
+		while (ring_buf_space_get(&ppp->transmit_rb) > 0) {
+			byte = modem_ppp_wrap_net_pkt_byte(ppp);
+
+			ring_buf_put(&ppp->transmit_rb, &byte, 1);
+
+			if (ppp->transmit_state == MODEM_PPP_TRANSMIT_STATE_IDLE) {
+				net_pkt_unref(ppp->tx_pkt);
+
+				modem_ppp_tx_net_pkt_buf_get(ppp, &ppp->tx_pkt);
+
+				break;
+			}
+		}
+	}
+
+	reserved_size = ring_buf_get_claim(&ppp->transmit_rb, &reserved, UINT32_MAX);
+
+	if (reserved_size == 0) {
+		ring_buf_get_finish(&ppp->transmit_rb, 0);
+
+		return;
+	}
+
+	ret = modem_pipe_transmit(ppp->pipe, reserved, reserved_size);
+
+	if (ret < 0) {
+		ring_buf_get_finish(&ppp->transmit_rb, 0);
+	} else {
+		ring_buf_get_finish(&ppp->transmit_rb, (uint32_t)ret);
+	}
+
+	/* Resubmit send work if data remains */
+	if ((ring_buf_is_empty(&ppp->transmit_rb) == false) || (ppp->tx_pkt != NULL)) {
+		k_work_submit(&ppp->send_work.work);
+	}
+}
+
+static void modem_ppp_process_handler(struct k_work *item)
+{
+	struct modem_ppp_work_item *ppp_work_item = (struct modem_ppp_work_item *)item;
+	struct modem_ppp *ppp = ppp_work_item->ppp;
+	int ret;
+
+	ret = modem_pipe_receive(ppp->pipe, ppp->receive_buf, ppp->buf_size);
+
+	if (ret < 1) {
+		return;
+	}
+
+	for (int i = 0; i < ret; i++) {
+		modem_ppp_process_received_byte(ppp, ppp->receive_buf[i]);
+	}
+
+	k_work_submit(&ppp->process_work.work);
+}
+
+static void modem_ppp_ppp_api_init(struct net_if *iface)
+{
+	const struct device *dev = net_if_get_device(iface);
+
+	struct modem_ppp *ppp = (struct modem_ppp *)dev->data;
+
+	net_ppp_init(iface);
+
+	net_if_flag_set(iface, NET_IF_NO_AUTO_START);
+
+	if (ppp->init_iface != NULL) {
+		ppp->init_iface(iface);
+	}
+
+	ppp->iface = iface;
+}
+
+static int modem_ppp_ppp_api_start(const struct device *dev)
+{
+	return 0;
+}
+
+static int modem_ppp_ppp_api_stop(const struct device *dev)
+{
+	return 0;
+}
+
+static int modem_ppp_ppp_api_send(const struct device *dev, struct net_pkt *pkt)
+{
+	struct modem_ppp *ppp = (struct modem_ppp *)dev->data;
+
+	if (atomic_test_bit(&ppp->state, MODEM_PPP_STATE_ATTACHED_BIT) == false) {
+		return -EPERM;
+	}
+
+	/* Validate packet protocol */
+	if ((net_pkt_is_ppp(pkt) == false) && (net_pkt_family(pkt) != AF_INET) &&
+	    (net_pkt_family(pkt) != AF_INET6)) {
+		return -EPROTONOSUPPORT;
+	}
+
+	/* Validate packet data length */
+	if (((net_pkt_get_len(pkt) < 2) && (net_pkt_is_ppp(pkt) == true)) ||
+	    ((net_pkt_get_len(pkt) < 1))) {
+		return -ENODATA;
+	}
+
+	net_pkt_ref(pkt);
+
+	if (modem_ppp_tx_net_pkt_buf_put(ppp, pkt) == false) {
+		LOG_WRN("tx pkt buf overrun");
+
+		net_pkt_unref(pkt);
+
+		return -ENOMEM;
+	}
+
+	k_work_submit(&ppp->send_work.work);
+
+	return 0;
+}
+
+const struct ppp_api modem_ppp_ppp_api = {
```

Comment:
I'm puzzled about the relationship with `drivers/net/ppp.c` driver. It seems a bit off to have pretty similar functionality, implementing the same API, in two different places. Shouldn't the old driver be updated instead? Or depracated or something?

Reply:
One goal of this PR is to act as a replacement the existing `gsm_ppp.c`, `drivers/net/ppp.c`, `drivers/console/gsm_mux.c` and `drivers/console/uart_mux.c` and accompanying KConfigs. If developers gravitate towards using the driver and modules in this PR, we can deprecate the existing subsystem in the future, however they can both happily coexist in the meanwhile